tests/components/samsungtv/test_config_flow.py

```diff
@@ -0,0 +1,388 @@
+        )
+        assert result["type"] == "form"
+        assert result["step_id"] == "confirm"
+
+        # device not found
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"], user_input="whatever"
+        )
+        assert result["type"] == "abort"
+        assert result["reason"] == "not_found"
+
+
+async def test_ssdp_already_in_progress(hass, remote):
+    """Test starting a flow from discovery twice."""
+
+    # confirm to add the entry
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": "ssdp"}, data=MOCK_SSDP_DATA
+    )
+    assert result["type"] == "form"
+    assert result["step_id"] == "confirm"
+
+    # failed as already in progress
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": "ssdp"}, data=MOCK_SSDP_DATA
+    )
+    assert result["type"] == "abort"
+    assert result["reason"] == "already_in_progress"
+
+
+async def test_ssdp_already_configured(hass, remote):
+    """Test starting a flow from discovery when already configured."""
+
+    # entry was added
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": "user"}, data=MOCK_USER_DATA
+    )
+    assert result["type"] == "create_entry"
+    assert result["data"][CONF_MANUFACTURER] is None
+    assert result["data"][CONF_MODEL] is None
+    assert result["data"][CONF_ID] is None
+
+    # failed as already configured
+    result2 = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": "ssdp"}, data=MOCK_SSDP_DATA
+    )
+    assert result2["type"] == "abort"
+    assert result2["reason"] == "already_configured"
+
+    # check updated device info
+    assert result["data"][CONF_MANUFACTURER] == "fake_manufacturer"
+    assert result["data"][CONF_MODEL] == "fake_model"
+    assert result["data"][CONF_ID] == "fake_uuid"
+
+
+async def test_autodetect_websocket(hass, remote):
+    """Test for send key with autodetection of protocol."""
+    with patch("homeassistant.components.samsungtv.config_flow.Remote") as remote:
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": "user"}, data=MOCK_USER_DATA
+        )
+        assert result["type"] == "create_entry"
+        assert result["data"][CONF_METHOD] == "websocket"
+        assert remote.call_count == 1
+        assert remote.call_args_list == [call(AUTODETECT_WEBSOCKET)]
+
+
+async def test_autodetect_auth_missing(hass, remote):
+    """Test for send key with autodetection of protocol."""
+    with patch(
+        "homeassistant.components.samsungtv.config_flow.Remote",
+        side_effect=[AccessDenied("Boom")],
+    ) as remote:
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": "user"}, data=MOCK_USER_DATA
+        )
+        assert result["type"] == "abort"
+        assert result["reason"] == "auth_missing"
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=4795320)